### PR TITLE
Replace connection in bucket constructor.

### PR DESCRIPTION
**Pull request by dhermes**
Has #984 as diffbase.

Also updating storage.Batch to accept a client.

Introduced temporary hack to allow system-tests to continue passing. Implicit environment has still not been completely removed.

### CODE REVIEW

File: gcloud/storage/client.py [hunk 1]
```diff
@@ -205,6 +205,6 @@ def get_items_from_response(self, response):
         """
         for item in response.get('items', []):
             name = item.get('name')
-            bucket = Bucket(name)
+            bucket = Bucket(None, name)
```
Comment (dhermes) [inline gcloud/storage/client.py]: Old comments from 8b1ecb201cb63d46bea2fc79c28dec6d257c7d24

> @tseaver Why create a Bucket w/o a client? Its methods won't be usable.

> @dhermes: Forthcoming in ensuing PRs.

File: system_tests/storage.py [hunk 2]
```diff
@@ -51,9 +51,15 @@ def setUp(self):
...
+        with storage.Batch(CLIENT) as batch:
+            # Stop-gap measure to support batches during transation to
+            # to clients from implicit behavior.
+            batch_client = storage.Client(
+                project=CLIENT.project,
+                credentials=CLIENT.connection.credentials)
+            batch_client.connection = batch
             for bucket_name in self.case_buckets_to_delete:
-                storage.Bucket(bucket_name).delete()
+                storage.Bucket(batch_client, name=bucket_name).delete()
```
Comment (dhermes) [inline system_tests/storage.py]: Old comments from 8b1ecb201cb63d46bea2fc79c28dec6d257c7d24

> @tseaver: Ugh

> @dhermes: It's fine. It'll go away by the end of the PRs. I just didn't want to do everything all at once.

**Comment by dhermes** [comment]
@tseaver Any remaining issues?

File: gcloud/storage/__init__.py [hunk 3]
```diff
@@ -78,7 +78,7 @@ def set_default_bucket(bucket=None):
         bucket_name = os.getenv(_BUCKET_ENV_VAR_NAME)
 
         if bucket_name is not None:
-            bucket = Bucket(bucket_name)
+            bucket = Bucket(None, name=bucket_name)
```
Comment (tseaver) [inline gcloud/storage/__init__.py]: I see that #989 removes ``storage.set_default_bucket`` altogether.

File: gcloud/storage/bucket.py [hunk 4]
```diff
@@ -205,6 +205,11 @@ def path(self):
 
         return self.path_helper(self.name)
 
+    @property
```
Comment (tseaver) [inline gcloud/storage/bucket.py]: I've got a sense that this property (maybe the others too) should be moved up to before the methods which use it (at least, I was concerned about the `_client` vs. `client` usage in this diff).

Comment (dhermes) [inline gcloud/storage/bucket.py]: I'm not sure what you mean by before? You want me to re-write all the commits or move it up vertically in a file? Also, all the PRs strung together bring us to a "final state" and most states in between are just transient.

File: gcloud/storage/client.py (hunk 1 above)
Comment (tseaver) [inline gcloud/storage/client.py]: Looks like #989 addresses this.

File: gcloud/storage/bucket.py (hunk 4 above)
Comment (tseaver) [inline gcloud/storage/bucket.py]: We can move it later, but having methods use `self.client` when the ctor only assigns `self._client` and the property definintion is buried hundreds of lines down is a readability issue.

**Comment by tseaver** [comment]
We can merge this one to get to the next.

File: gcloud/storage/bucket.py (hunk 4 above)
Comment (dhermes) [inline gcloud/storage/bucket.py]: Gotcher. I'm happy to move it vertically. #997 filed

**Pull request merged**